### ISSUE TITLE
Model only has gradients when "x = x + bias" and not when "x += bias"

### ISSUE THREAD

**Issue by mxbi** [pytorch/pytorch#9634]
## Issue description

I've built a simple model with a single bias layer (adding a set of constants to the inputs) as follows:

```python
class Bias(nn.Module):
    def __init__(self, dim):
        super(Bias, self).__init__()
        self.bias = nn.Parameter(torch.zeros(dim))

    def forward(self, x):
        x += self.bias
        return x
```
However, when I try training with this with an Adam optimizer, the weights of the bias layer do not change at all. Checking `list(model.parameters())` shows that it is indeed in the optimizer, and checking `list(model.parameters()[0].grad)` after running `loss.backward()` returns a vector of all zeros.

As soon as I change the `x += self.bias` to `x = x + self.bias` in the forward() function, the model trains perfectly. The same issue also occurs with other operators such as `*=`. This model works:

```python
class Bias(nn.Module):
    def __init__(self, dim):
        super(Bias, self).__init__()
        self.bias = nn.Parameter(torch.zeros(dim))

    def forward(self, x):
        x = x + self.bias
        return x
```
It seems there is some difference with how the gradients of these are calculated which means that the gradients are all-zero when += is used and so training the model does not work. Is this intended behaviour?

## System Info
```
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 8.0.61

OS: Ubuntu 16.04.2 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.4) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 8.0.61
GPU models and configuration: 
GPU 0: TITAN X (Pascal)

Nvidia driver version: 384.81
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy (1.14.5)
[pip] numpydoc (0.8.0)
[pip] torch (0.4.0)
[pip] torchfile (0.1.0)
[pip] torchsummary (1.3)
[pip] torchvision (0.2.1)
[conda] pytorch                   0.4.0           py36_cuda8.0.61_cudnn7.1.2_1    pytorch
[conda] torchfile                 0.1.0                     <pip>
[conda] torchsummary              1.3                       <pip>
[conda] torchvision               0.2.1                    py36_1    pytorch
```

**Comment by t-vi** [comment]
Yes, this is intended behaviour.
The `=+, =*, ...` ops are inplace operations (like add_) and you cannot keep backward through them.
When using `x = x + y` the old `x` stays around (but doesn't have a name anymore) and will be visited in the backward.

**Comment by mxbi** [comment]
Thanks, I will watch out for this in the future.

**Comment by SsnL** [comment]
The above explanation is misleading. One should be able to backprop through inplace ops. The following script works correctly on a master build:
```py
import torch
import torch.nn as nn

class Bias(nn.Module):
    def __init__(self, dim):
        super(Bias, self).__init__()
        self.bias = nn.Parameter(torch.zeros(dim))

    def forward(self, x):
        x += self.bias
        return x

a = torch.randn(1)
b = Bias(1)

l = b(a)
l.backward()
print(b.bias.grad)  # [1.]
```
